### ISSUE TITLE
Different meaning of 'arg: MyClass' upon --legacy-nilable-classes

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#13720]
`--legacy-nilable-classes` (on by default) has the effect of treating a formal written `arg: MyClass` as having the borrowed MyClass type. Which is not generic if MyClass is not.

Cf. with `--no-legacy-nilable-classes` the formal's type is the generic any-managed MyClass, as it is uspposed to be.

The formal's type is non-nilable in both cases, which is correct.

For example, the following program produces different compiler output depending on the legacy flag:

```chpl
class Test { }

proc testme(arg: Test) {
  compilerError(arg.type:string);
}

testme(new unmanaged Test());
```

**Comment by mppf** [comment]
@vasslitvinov - I might be missing something but the flag seems to be operating exactly as expected. Why is this an issue?

**Comment by vasslitvinov** [comment]
> the flag seems to be operating exactly as expected

Maybe my expectations are off? The option is documented as `Allow all variables of class type to store nil`, maybe it should be expanded to include the "borrow vs. generic management" distinction?

Also, --no-legacy-nilable-classes messes PRIM_GATHER_TESTS and makes this test fail:

    library/packages/UnitTest/PopulateTests_PRIM/TestCount

In detail, PRIM_GATHER_TESTS looks for functions that are not generic. `proc test_square(test: Test)` is generic, so it is not included in the set of tests. Hm, how does it work in the rest of the test suite?

**Comment by mppf** [comment]
> maybe it should be expanded to include the "borrow vs. generic management" distinction?

Yes and we additionally probably need to rename the flag to `--legacy-classes`.

> In detail, PRIM_GATHER_TESTS looks for functions that are not generic. proc test_square(test: Test) is generic, so it is not included in the set of tests. Hm, how does it work in the rest of the test suite?

I updated the other tests to be written as `proc someTest(test: borrowed Test)`.

However I think ultimately PRIM_GATHER_TESTS should be updated to gather instantiations with a `borrowed Test`.

**Comment by vasslitvinov** [comment]
+1 to `--legacy-classes`. I suspect its help string will be on the longer side.

Meanwhile #13755 talks specifically about PRIM_GATHER_TESTS.
